week3/d01/Assignment_3_1_1/ex3.cpp

```diff
@@ -0,0 +1,88 @@
+#include <iostream>
+#include <string>
+#include <vector>
+#include <limits>
+#include <exception>
+
+void clear_cin()
+{
+    std::cin.clear();
+    std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
+}
+
+struct Names
+{
+    int id {};
+    std::string name {};
+};
+
+void print_names(std::vector<Names>& names)
+{
+    for (Names name : names)
+    {
+        std::cout << name.id << ") " << name.name << '\n';
+    }
+}
+
+/*
+**  Asks user for an id,. Keeps asking if user inputs something that will fail
+**  std::cin, like characters or max int/min int etc.
+*/
+int ask_id()
+{
+    while (true)
+    {
+        std::cout << "Select an id: ";
+        int id {};
+        std::cin >> id;
+        if (!std::cin)
+            clear_cin();
+        else
+            return id;
+    }
+}
+
+void check_names(std::vector<Names>& names, int id)
+{
+    for (Names name : names)
+    {
+        if (name.id == id)
+        {
+            std::cout << "\nYou selected " << name.name << '\n';
+            return ;
+        }
+    }
+    throw std::range_error("Selected id is out of range!");
```

Comment:
I think this could be thrown as something more general, as I think range_error isn't usually used in this way:


["Defines a type of object to be thrown as exception. It can be used to report range errors (that is, situations where a result of a computation cannot be represented by the destination type)."](https://en.cppreference.com/w/cpp/error/range_error)